core-bundle/contao/drivers/DC_Table.php

```diff
@@ -4619,16 +4618,39 @@ protected function parentView()
-<a href="' . preg_replace('/&(amp;)?table=[^& ]*/i', $this->ptable ? '&amp;table=' . $this->ptable : '', $this->addToUrl('act=edit' . (Input::get('nb') ? '&amp;nc=1' : ''))) . '" class="edit" title="' . StringUtil::specialchars(\sprintf(\is_array($labelEditHeader) ? $labelEditHeader[1] : $labelEditHeader, $objParent->id)) . '">' . $imageEditHeader . '</a> ' . $this->generateHeaderButtons($objParent->row(), $this->ptable) : '') . (($blnHasSorting && !($GLOBALS['TL_DCA'][$this->strTable]['config']['closed'] ?? null) && !($GLOBALS['TL_DCA'][$this->strTable]['config']['notCreatable'] ?? null) && $security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, $this->addDynamicPtable(array('pid' => $objParent->id, 'sorting' => 0))))) ? '
-<a href="' . $this->addToUrl('act=create&amp;mode=2&amp;pid=' . $objParent->id . '&amp;id=' . $this->intId) . '" title="' . StringUtil::specialchars($labelPasteNew[0]) . '">' . $imagePasteNew . '</a>' : ''))) . '
+<div class="tl_content_right">';
+
+			if (Input::get('act') == 'select' || $this->strPickerFieldType == 'checkbox')
+			{
+				$return .= '
+<label for="tl_select_trigger" class="tl_select_label">' . $GLOBALS['TL_LANG']['MSC']['selectAll'] . '</label> <input type="checkbox" id="tl_select_trigger" onclick="Backend.toggleCheckboxes(this)" class="tl_tree_checkbox">';
+			}
+			else
+			{
+				if ($blnClipboard)
+				{
+					$return .= '
+<a href="' . $this->addToUrl('act=' . $arrClipboard['mode'] . '&amp;mode=2&amp;pid=' . $objParent->id . (!$blnMultiboard ? '&amp;id=' . $arrClipboard['id'] : '')) . '" title="' . StringUtil::specialchars($labelPasteAfter[0]) . '" data-action="contao--scroll-offset#store">' . $imagePasteAfter . '</a>';
+				}
+				else
+				{
+					$buttons = $this->generateHeaderButtons($objParent->row(), $this->ptable);
+
+					if ($blnHasSorting && !($GLOBALS['TL_DCA'][$this->strTable]['config']['closed'] ?? null) && !($GLOBALS['TL_DCA'][$this->strTable]['config']['notCreatable'] ?? null) && $security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, $this->addDynamicPtable(array('pid' => $objParent->id, 'sorting' => 0)))))
+					{
+						$create = '<a href="' . $this->addToUrl('act=create&amp;mode=2&amp;pid=' . $objParent->id . '&amp;id=' . $this->intId) . '" title="' . StringUtil::specialchars($labelPasteNew[0]) . '">' . $imagePasteNew . '</a>';
+						$buttons->append(array('html' => $create, 'primary' => true));
```

Comment:
<img width="727" alt="" src="https://github.com/user-attachments/assets/a244969d-db09-4d4f-bfff-5b935a65fabe">

Reply:
yes, that's something I asked for at the dev meeting. There's currently no syntax to tell phpstan/IDEs about the dynamic class we are creating here. The only alternative would be to create three class files for this minor implementation and use a factory to create the builder. But I kinda tried to refrain from that _overkill_ 🙃

Reply:
Also, I'd love to find a better way to implement this with more refactoring, but that's the best I could come up for now. Be aware that the builder is `@internal` so we can always change that.

Reply:
The return type of the `generatedHeaderButtons()` method is `string`. If we change it accordingly, PhpStorm should understand, shouldn‘t it?

Reply:
you mean the one in the annotation? I'm not sure PHPStorm will, because we return a dynamic object without any interface.

Reply:
I think we should enhance the `generateHeaderButtons()` method to accept a third parameter:

```php
protected function generateHeaderButtons($arrRow, $strPtable, array $additionalOperations = array())
```

Then we can always return a string in the `generateHeaderButtons()` method and don‘t need a dynamic object.

Reply:
changed in 383a343724428e5da5a72a77f42ecb5d746e805e